File: src/Client.Core/Decoding/Fetcher.cs
namespace Client.Core.Decoding;

public sealed class Fetcher
{
  private readonly List<GopEntry> _gops = [];
  private readonly List<(ulong From, ulong To)> _gaps = [];
  private readonly Lock _lock = new();

  private Func<ulong, ulong, Task>? _sendFetch;
  private bool _fetchInFlight;
  private bool _live;
  private TaskCompletionSource? _fetchTcs;

  public int BufferedGopCount
  {
    get { lock (_lock) return _gops.Count; }
  }

  public int BufferedBytes
  {
    get
    {
      lock (_lock)
      {
        var total = 0;
        foreach (var g in _gops)
          foreach (var c in g.Chunks) total += c.Length;
        return total;
      }
    }
  }

  public void Attach(Func<ulong, ulong, Task> sendFetch)
  {
    _sendFetch = sendFetch;
  }

  public void Detach()
  {
    _sendFetch = null;
  }

  public void SetTarget(long fromUs, long toUs)
  {
    var forward = toUs > fromUs;

    lock (_lock)
    {
      if (_gops.Count > 1)
      {
        var containing = FindGopLocked(fromUs);
        if (containing != null)
        {
          var idx = _gops.IndexOf(containing);
          if (forward && idx > 1)
            _gops.RemoveRange(0, idx - 1);
          else if (!forward && idx < _gops.Count - 2)
            _gops.RemoveRange(idx + 2, _gops.Count - (idx + 2));
        }
      }

      if (_fetchInFlight || _sendFetch == null || _live) return;

      ulong? fetchFrom = null;
      ulong fetchTo = 0;

      if (forward)
      {
        var newest = NewestLocked();
        if (newest == null || (long)newest.Value < toUs)
        {
          var from = newest != null ? newest.Value + 1 : (ulong)fromUs;
          from = SkipGapsForwardLocked(from);
          fetchFrom = from;
          fetchTo = from + 30_000_000;
        }
      }
      else
      {
        var oldest = OldestLocked();
        if (oldest == null || (long)oldest.Value > toUs)
        {
          var from = oldest != null ? oldest.Value - 1 : (ulong)fromUs;
          from = SkipGapsReverseLocked(from);
          fetchFrom = from;
          fetchTo = from >= 30_000_000 ? from - 30_000_000 : 0;
        }
      }

      if (fetchFrom == null) return;
      _fetchInFlight = true;
      _ = _sendFetch(fetchFrom.Value, fetchTo);
    }
  }

  public void HandleFetchComplete()
  {
    TaskCompletionSource? tcs;
    lock (_lock)
    {
      _fetchInFlight = false;
      tcs = _fetchTcs;
      _fetchTcs = null;
    }
    tcs?.TrySetResult();
  }

  public Task FetchAtAsync(ulong ts)
  {
    lock (_lock)
    {
      if (_sendFetch == null) return Task.CompletedTask;
      if (_fetchInFlight)
      {
        var existing = _fetchTcs ??= new TaskCompletionSource();
        return existing.Task;
      }
      _fetchInFlight = true;
      _fetchTcs = new TaskCompletionSource();
      var result = _fetchTcs.Task;
      _ = _sendFetch(ts, ts);
      return result;
    }
  }

  public void HandleLive() { lock (_lock) _live = true; }
  public void HandleRecording() { lock (_lock) _live = false; }

  public void HandleGap(ulong from, ulong to)
  {
    lock (_lock) _gaps.Add((from, to));
  }

  public void AppendData(ulong timestamp, ReadOnlyMemory<byte> chunk)
  {
    lock (_lock)
    {
      var existing = _gops.Find(g => g.Timestamp == timestamp);
      if (existing != null)
      {
        existing.Chunks.Add(chunk);
        return;
      }
      var entry = new GopEntry(timestamp);
      entry.Chunks.Add(chunk);
      _gops.Add(entry);
      _gops.Sort((a, b) => a.Timestamp.CompareTo(b.Timestamp));
    }
  }

  public GopEntry? FindGop(ulong timestamp)
  {
    lock (_lock) return FindGopLocked((long)timestamp);
  }

  public ulong[] GopTimestamps()
  {
    lock (_lock)
    {
      var result = new ulong[_gops.Count];
      for (var i = 0; i < _gops.Count; i++) result[i] = _gops[i].Timestamp;
      return result;
    }
  }

  public ulong? OldestTimestamp()
  {
    lock (_lock) return OldestLocked();
  }

  public ulong? NewestTimestamp()
  {
    lock (_lock) return NewestLocked();
  }

  public void Reset()
  {
    lock (_lock)
    {
      _gops.Clear();
      _gaps.Clear();
      _fetchInFlight = false;
      _live = false;
      _fetchTcs?.TrySetResult();
      _fetchTcs = null;
    }
  }

  public static ReadOnlyMemory<byte> MergedData(GopEntry gop)
  {
    if (gop.Chunks.Count == 1) return gop.Chunks[0];
    var total = 0;
    foreach (var c in gop.Chunks) total += c.Length;
    var merged = new byte[total];
    var off = 0;
    foreach (var c in gop.Chunks) { c.Span.CopyTo(merged.AsSpan(off)); off += c.Length; }
    return merged;
  }

  private GopEntry? FindGopLocked(long timestamp)
  {
    if (_gops.Count == 0) return null;
    var lo = 0;
    var hi = _gops.Count - 1;
    while (lo < hi)
    {
      var mid = (lo + hi + 1) >>> 1;
      if ((long)_gops[mid].Timestamp <= timestamp) lo = mid;
      else hi = mid - 1;
    }
    return (long)_gops[lo].Timestamp <= timestamp ? _gops[lo] : null;
  }

  private ulong? OldestLocked() => _gops.Count > 0 ? _gops[0].Timestamp : null;
  private ulong? NewestLocked() => _gops.Count > 0 ? _gops[^1].Timestamp : null;

  private ulong SkipGapsForwardLocked(ulong ts)
  {
    foreach (var gap in _gaps)
      if (ts >= gap.From && ts < gap.To) ts = gap.To;
    return ts;
  }

  private ulong SkipGapsReverseLocked(ulong ts)
  {
    foreach (var gap in _gaps)
      if (ts > gap.From && ts <= gap.To) ts = gap.From;
    return ts;
  }
}

public sealed class GopEntry(ulong timestamp)
{
  public ulong Timestamp { get; } = timestamp;
  public List<ReadOnlyMemory<byte>> Chunks { get; } = [];
}

File: src/Client.Web/src/media/fetcher.ts
export interface GopEntry {
  timestamp: number
  chunks: Uint8Array[]
}

export class Fetcher {
  private gops: GopEntry[] = []
  private gaps: { from: number, to: number }[] = []
  private fetchInFlight = false
  private live = false
  private lastFetchFrom: number | null = null
  private fetchProgressed = false
  private exhaustedFrom: number | null = null
  private exhaustedAt = 0
  private static readonly exhaustedRetryMs = 2000
  private sendFetch: ((from: number, to: number) => void) | null = null
  private fetchResolve: (() => void) | null = null

  attach(sendFetch: (from: number, to: number) => void) {
    this.sendFetch = sendFetch
  }

  detach() {
    this.sendFetch = null
  }

  setTarget(fromUs: number, toUs: number) {
    const forward = toUs > fromUs

    if (this.gops.length > 1) {
      const containing = this.findGop(fromUs)
      if (containing) {
        const idx = this.gops.indexOf(containing)
        if (forward && idx > 1)
          this.gops.splice(0, idx - 1)
        else if (!forward && idx < this.gops.length - 2)
          this.gops.splice(idx + 2)
      }
    }

    if (this.fetchInFlight || !this.sendFetch || this.live) return

    if (forward) {
      const newest = this.newestTimestamp()
      if (newest === null || newest < toUs) {
        let from = newest !== null ? newest + 1 : fromUs
        from = this.skipGapsForward(from)
        if (this.isExhausted(from)) return
        this.beginFetch(from)
        this.sendFetch(from, from + 30_000_000)
      }
    } else {
      const oldest = this.oldestTimestamp()
      if (oldest === null || oldest > toUs) {
        let from = oldest !== null ? oldest - 1 : fromUs
        from = this.skipGapsReverse(from)
        if (this.isExhausted(from)) return
        this.beginFetch(from)
        this.sendFetch(from, from - 30_000_000)
      }
    }
  }

  private beginFetch(from: number) {
    this.fetchInFlight = true
    this.lastFetchFrom = from
    this.fetchProgressed = false
  }

  private isExhausted(from: number): boolean {
    return from === this.exhaustedFrom
      && Date.now() - this.exhaustedAt < Fetcher.exhaustedRetryMs
  }

  handleFetchComplete() {
    if (this.lastFetchFrom !== null && !this.fetchProgressed) {
      this.exhaustedFrom = this.lastFetchFrom
      this.exhaustedAt = Date.now()
    }
    this.lastFetchFrom = null
    this.fetchInFlight = false
    if (this.fetchResolve) {
      this.fetchResolve()
      this.fetchResolve = null
    }
  }

  fetchAt(ts: number): Promise<void> {
    if (!this.sendFetch) return Promise.resolve()
    if (this.fetchInFlight) {
      return new Promise(resolve => {
        const prev = this.fetchResolve
        this.fetchResolve = () => { prev?.(); resolve() }
      })
    }
    this.fetchInFlight = true
    this.lastFetchFrom = null
    this.sendFetch(ts, ts)
    return new Promise(resolve => { this.fetchResolve = resolve })
  }

  stats(): { gops: number, bytes: number } {
    let bytes = 0
    for (const gop of this.gops)
      for (const chunk of gop.chunks)
        bytes += chunk.length
    return { gops: this.gops.length, bytes }
  }

  handleLive() {
    this.live = true
  }

  handleRecording() {
    this.live = false
  }

  handleGap(from: number, to: number) {
    this.gaps.push({ from, to })
    this.exhaustedFrom = null
  }

  appendData(timestamp: number, data: Uint8Array, begin: boolean) {
    const existing = this.gops.find(g => g.timestamp === timestamp)
    if (existing) {
      if (begin)
        existing.chunks = [data]
      else
        existing.chunks.push(data)
      return
    }
    this.fetchProgressed = true
    this.exhaustedFrom = null
    this.gops.push({ timestamp, chunks: [data] })
    this.gops.sort((a, b) => a.timestamp - b.timestamp)
  }

  findGop(timestamp: number): GopEntry | null {
    if (this.gops.length === 0) return null
    let lo = 0
    let hi = this.gops.length - 1
    while (lo < hi) {
      const mid = (lo + hi + 1) >>> 1
      if (this.gops[mid].timestamp <= timestamp)
        lo = mid
      else
        hi = mid - 1
    }
    return this.gops[lo].timestamp <= timestamp ? this.gops[lo] : null
  }

  mergedData(gop: GopEntry): Uint8Array {
    if (gop.chunks.length === 1) return gop.chunks[0]
    const totalLen = gop.chunks.reduce((sum, c) => sum + c.length, 0)
    const merged = new Uint8Array(totalLen)
    let offset = 0
    for (const chunk of gop.chunks) {
      merged.set(chunk, offset)
      offset += chunk.length
    }
    return merged
  }

  gopTimestamps(): number[] {
    return this.gops.map(g => g.timestamp)
  }

  oldestTimestamp(): number | null {
    return this.gops.length > 0 ? this.gops[0].timestamp : null
  }

  newestTimestamp(): number | null {
    return this.gops.length > 0 ? this.gops[this.gops.length - 1].timestamp : null
  }

  reset() {
    this.gops = []
    this.gaps = []
    this.fetchInFlight = false
    this.live = false
    this.lastFetchFrom = null
    this.fetchProgressed = false
    this.exhaustedFrom = null
  }

  private skipGapsForward(ts: number): number {
    for (const gap of this.gaps) {
      if (ts >= gap.from && ts < gap.to)
        ts = gap.to
    }
    return ts
  }

  private skipGapsReverse(ts: number): number {
    for (const gap of this.gaps) {
      if (ts > gap.from && ts <= gap.to)
        ts = gap.from
    }
    return ts
  }
}
